app/src/test/java/com/github/se/model/tags/TagsRepositoryTest.kt

```diff
@@ -0,0 +1,415 @@
+import com.github.se.icebreakrr.model.tags.TagsRepository
+import com.google.android.gms.tasks.OnFailureListener
+import com.google.android.gms.tasks.OnSuccessListener
+import com.google.android.gms.tasks.Task
+import com.google.firebase.firestore.CollectionReference
+import com.google.firebase.firestore.DocumentReference
+import com.google.firebase.firestore.DocumentSnapshot
+import com.google.firebase.firestore.FirebaseFirestore
+import com.google.firebase.firestore.QuerySnapshot
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertTrue
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.mockito.ArgumentMatchers.any
+import org.mockito.ArgumentMatchers.anyList
+import org.mockito.ArgumentMatchers.eq
+import org.mockito.Mockito.doAnswer
+import org.mockito.Mockito.mock
+import org.mockito.Mockito.spy
+import org.mockito.Mockito.`when`
+import org.mockito.invocation.InvocationOnMock
+import org.mockito.kotlin.anyOrNull
+import org.robolectric.RobolectricTestRunner
+
+@RunWith(RobolectricTestRunner::class)
+class TagsRepositoryTest {
+  private lateinit var firestore: FirebaseFirestore
+  private lateinit var repository: TagsRepository
+
+  @Before
+  fun setUp() {
+    firestore = mock(FirebaseFirestore::class.java)
+    repository = spy(TagsRepository(firestore))
+  }
+
+  @Test
+  fun getAllTagsOnSuccessTest() {
+    val repoAllTags: List<TagsCategory> =
+        listOf(
+            TagsCategory("Sport", "0xFFFF0000", listOf("Tennis", "Basketball", "PingPong")),
+            TagsCategory("Music", "0xFF0000FF", listOf("Rock", "Tech", "Classical")))
+
+    val collectionReference = mock(CollectionReference::class.java)
+    val task = mock(Task::class.java) as Task<QuerySnapshot>
+    val querySnapshot = mock(QuerySnapshot::class.java)
+    val doc1 = mock(DocumentSnapshot::class.java)
+    val doc2 = mock(DocumentSnapshot::class.java)
+    val documents = listOf(doc1, doc2)
+
+    `when`(firestore.collection("Tags")).thenReturn(collectionReference)
+    `when`(collectionReference.get()).thenReturn(task)
+    `when`(querySnapshot.documents).thenReturn(documents)
+    `when`(doc1.exists()).thenReturn(true)
+    `when`(doc2.exists()).thenReturn(true)
+
+    `when`(doc1.getString("name")).thenReturn("Sport")
+    `when`(doc1.getString("color")).thenReturn("0xFFFF0000")
+    `when`(doc1.get("subtags")).thenReturn(listOf("Tennis", "Basketball", "PingPong"))
+
+    `when`(doc2.getString("name")).thenReturn("Music")
+    `when`(doc2.getString("color")).thenReturn("0xFF0000FF")
+    `when`(doc2.get("subtags")).thenReturn(listOf("Rock", "Tech", "Classical"))
+
+    `when`(task.addOnSuccessListener(any())).thenAnswer { invocation: InvocationOnMock ->
+      val listener = invocation.arguments[0] as OnSuccessListener<QuerySnapshot>
+      listener.onSuccess(querySnapshot)
+      task
+    }
+    `when`(task.addOnFailureListener(any())).thenAnswer { invocation: InvocationOnMock -> task }
```

Comment:
Problem (ft. copilot): it seems there is a chaining problem in that mocking. As I'm getting really tired I'll let you do the research but here is the copilot message (maybe it is completely bogus): 

> Incorrect Mocking of Task
The issue here is with the mocking of the addOnSuccessListener and addOnFailureListener methods for the Task object. These methods are mocked but not properly chained, which might lead to unexpected behavior. Specifically, the addOnSuccessListener method should return the Task object to allow method chaining, but this is not guaranteed in the provided code.

Reply:
I thing copilot is completely bogus (I indeed return the task).